rust/core-lib/src/tabs.rs

```diff
@@ -18,53 +18,183 @@ use std::collections::BTreeMap;
 use std::io::{self, Read, Write};
 use std::path::{PathBuf, Path};
 use std::fs::File;
-use std::sync::{Arc, Mutex};
+use std::sync::{Arc, Weak};
+
 use serde_json::value::Value;
+use parking_lot::{MutexGuard, Mutex};
 
 use xi_rope::rope::Rope;
 use editor::Editor;
-use rpc::{CoreCommand, EditCommand};
+use rpc::{CoreCommand, EditCommand, PluginCommand};
 use styles::{Style, StyleMap};
 use MainPeer;
+use plugins::PluginManager;
 
 /// ViewIdentifiers are the primary means of routing messages between xi-core and a client view.
 pub type ViewIdentifier = String;
 
 /// BufferIdentifiers uniquely identify open buffers.
-type BufferIdentifier = String;
+pub type BufferIdentifier = String;
 
-// TODO: proposed new name: something like "Core" or "CoreState" or "EditorState"? "Documents?"
-pub struct Tabs<W: Write> {
-    /// maps file names to buffer identifiers. If a client asks to open a file that is already
-    /// open, we treat it as a request for a new view.
+
+/// Tracks open buffers, and relationships between buffers and views.
+pub struct BufferContainer<W: Write> {
+    /// associates open file paths to buffers
     open_files: BTreeMap<PathBuf, BufferIdentifier>,
     /// maps buffer identifiers (filenames) to editor instances
-    buffers: BTreeMap<BufferIdentifier, Arc<Mutex<Editor<W>>>>,
-    /// maps view identifiers to editor instances. All actions originate in a view; this lets us
-    /// route messages correctly when multiple views share a buffer.
+    editors: BTreeMap<BufferIdentifier, Editor<W>>,
+    /// maps view identifiers to buffer identifiers. All actions originate in a view;
+    /// this lets us route messages correctly when multiple views share a buffer.
     views: BTreeMap<ViewIdentifier, BufferIdentifier>,
+}
+
+/// Wrapper around `Arc<Mutex<`[`BufferContainer`][BufferContainer]`>>`,
+/// for more ergonomic synchronization.
+///
+/// `BufferContainerRef` provides a thread-safe API for accessing and modifying the
+/// [`BufferContainer`][BufferContainer]. All methods on `BufferContainerRef`
+/// are thread safe. For finer grained actions, the underlying container can be accessed
+/// via `BufferContainer::lock`.
+///
+/// [BufferContainer]: struct.BufferContainer.html
+pub struct BufferContainerRef<W: Write>(Arc<Mutex<BufferContainer<W>>>);
+
+/// Wrapper around a `Weak<Mutex<BufferContainer<W>>>`
+///
+/// `WeakBufferContaienrRef` provides a more ergonomic way of storing a `Weak`
+/// reference to a [`BufferContainer`][BufferContainer].
+///
+/// [BufferContainer]: struct.BufferContainer.html
+pub struct WeakBufferContainerRef<W: Write>(Weak<Mutex<BufferContainer<W>>>);
+
+impl <W:Write>BufferContainer<W> {
+    /// Returns a reference to the `Editor` instance owning `view_id`'s view.
+    ///
+    /// Panics if no buffer is associated with `view_id`.
```

Comment:
Agreed, there's a TODO to this effect in here somewhere. Basically I would like to do proper error handling here, but unwraps felt okay for this as a proof of concept.